### PR TITLE
reviews

### PR DESCRIPTION
**Pull request by ps789**

### CODE REVIEW

File: data_utils.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import numpy as np
+
+
+def compute_indices(n, train_prop: float, valid_prop: float):
```
Comment (ps789) [inline data_utils.py:4]: Why are we choosing to use compute_indices instead of a standard sklearn train_test_split? Took a look at the code using the method and it seems like the indices aren't used except for generating a split

File: dataset.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def load_dataset(dataset_name: str, train_prop: float, valid_prop: float) -> Dataset:
+    """
+    Main dataset loading function, which returns dataset in form of numpy arrays (X,y)
+    """
+    if dataset_name == 'taiwan_credit_risk':
+        dataset = load_taiwan_credit_risk(train_prop, valid_prop)
+    elif dataset_name == 'ames_housing':
+        assert train_prop + valid_prop == 1
```
Comment (ps789) [inline dataset.py:37]: Does this mean no test set? And if so, why only for ames_housing and not for any of the others?

Comment (ps789) [reply dataset.py:37]: Oh I see, separate test csv. Might want to note this down in a comment?

File: dataset.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+def load_civil_war_onset(train_prop, valid_prop) -> Dataset:
+    # Dataset from Ethnicity, Insurgency, and Civil War by Fearon and Laitin 2003
+    # Used by https://journals.sagepub.com/doi/pdf/10.1177/2053168020905487
+
+    df = pd.read_csv('datasets/fearson_laitin.csv')
+    X = df['onset']
+    y = df.loc[:, df.columns != 'onset']
+
+    raise NotImplementedError
```
Comment (ps789) [inline dataset.py:95]: oof

Comment (ps789) [reply dataset.py:95]: :)

File: gam.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
+# %%
+import numpy as np
+import matplotlib.pyplot as plt
+from models import *
+
+size = np.array([1.42,1.58,1.78,1.99,1.99,1.99,2.13,2.13,2.13,2.32,2.32,2.32,2.32,2.32,2.43,2.43,2.78,2.98,2.98])
```
Comment (ps789) [inline gam.py:6]: If testing using numbers from another source you'll probably want to cite that so people have context so like: https://reseau-mexico.fr/sites/reseau-mexico.fr/files/igam.pdf page 137

File: gam.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+wear = np.array([4.0,4.2,2.5,2.6,2.8,2.4,3.2,2.4,2.6,4.8,2.9,3.8,3.0,2.7,3.1,3.3,3.0,2.8,1.7])
```
Comment (ps789) [inline gam.py:7]: Same as above

File: gam.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+y = data[:,-1]
+# Rescale predictors onto [0,1]
+
+scaler_X = Scaler(X)
+
+scaler_y = Scaler(y)
+X_s = scaler_X.transform(X)
+y_s = scaler_y.transform(y)
+
+lmdas = [0.01024, 5368.70912]
```
Comment (ps789) [inline gam.py:57]: Same as previous, page 147. Want to mention how these were obtained (eg through gridsearch as in the paper)

File: models.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    res_below = data_below[:,-1] - below_mean 
+    res_above = data_above[:,-1] - above_mean
+    # Concatenating the residuals 
+    r = np.concatenate((res_below, res_above), axis=None)
+
+    # Calculating the mse 
+    n = len(r)
+    r = r ** 2
+    r = np.sum(r)
+    mse_split = r / n
```
Comment (ps789) [inline models.py:25]: why sum then divide instead  of just np.mean?

File: models.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    return mse_split
+
+def calc_mse_whole(data):
+    r = data[:,-1]-np.mean(data[:,-1])
+    n = len(r)
+    r = r ** 2
+    r = np.sum(r)
+    mse = r / n
```
Comment (ps789) [inline models.py:33]: same as above

File: models.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    def __init__(self, x, xk):
+        self.x = x
+        self.xk = xk # vector of knots
+        self.q = len(self.xk) + 2 # dimension of basis
+
+    @staticmethod
+    def R(x,z):
+        ''' R(x,z) for cubic spline on [0,1]'''
+        return ((z-0.5)**2-1/12) * ((x-0.5)**2-1/12)/4-((abs(x-z)-0.5)**4-(abs(x-z)-0.5)**2/2+7/240)/24
+
```
Comment (ps789) [inline models.py:123]: Also want to note that this is a very specific spline, so might want to denote either in the notes or class name or both.

File: dataset.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+        self.X_valid = X[val_indices]
+        self.y_valid = y[val_indices]
+
+        self.X_test = X[test_indices]
+        self.y_test = y[test_indices]
+
+
+def load_dataset(dataset_name: str, train_prop: float, valid_prop: float) -> Dataset:
+    """
```
Comment (ew423) [inline dataset.py:31]: Maybe tell people what train_prop and valid_prop are? Might be confusing for non-ML people.

File: data_utils.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    """ Compute the indices for train, val, test splits. """
+    train_num = int(n * train_prop)
```
Comment (WendyH1108) [inline data_utils.py:6]: int() method will always give the lower integer bound

**Review by ew423 — COMMENTED** [review]
Code review this time pretty much comes down to readability + documentation, esp when you have var names with a few letters - should probably clarify what they are in the method 

Also, saw a few chunks at the bottom (e.g. different Splines) that didn't have much explanation on them - might be useful to tell people what they do.

Overall good stuff, I'm just nitpicking a little, good work guys :)

File: dataset.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    return dataset
+
+
+def load_taiwan_credit_risk(train_prop, valid_prop) -> Dataset:
+    df = pd.read_csv('datasets/taiwan_credit_risk.csv')
+    X = df.iloc[:, :-1].to_numpy()
+    y = df.iloc[:, -1].to_numpy()
+
+    dataset = Dataset(X, y, train_prop, valid_prop)
+    return dataset
```
Comment (WendyH1108) [inline dataset.py:52]: Will it be better to write these preprocessing steps in different classes/files?

File: gam.py (hunk 4 above)
Comment (WendyH1108) [reply gam.py:6]: kinda confusing about the meaning of size here

File: models.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    def __init__(self, x, xk):
+        self.x = x
+        self.xk = xk # vector of knots
+        self.q = len(self.xk) + 2 # dimension of basis
+
+    @staticmethod
+    def R(x,z):
+        ''' R(x,z) for cubic spline on [0,1]'''
+        return ((z-0.5)**2-1/12) * ((x-0.5)**2-1/12)/4-((abs(x-z)-0.5)**4-(abs(x-z)-0.5)**2/2+7/240)/24
```
Comment (WendyH1108) [inline models.py:122]: Very specific functions, not sure if you have any plan to make this class and functions more general or denote by specific-named function.
